src/librbd/librbd.cc

```diff
@@ -2638,6 +2650,58 @@ extern "C" int rbd_mirror_peer_set_cluster(rados_ioctx_t p, const char *uuid,
   return librbd::api::Mirror<>::peer_set_cluster(io_ctx, uuid, cluster_name);
 }
 
+extern "C" int rbd_mirror_peer_get_attributes(
+    rados_ioctx_t p, const char *uuid, char *key_values, size_t *max_len) {
+  librados::IoCtx io_ctx;
+  librados::IoCtx::from_rados_ioctx_t(p, io_ctx);
+
+  std::map<std::string, std::string> attributes;
+  int r = librbd::api::Mirror<>::peer_get_attributes(io_ctx, uuid, &attributes);
+  if (r < 0) {
+    return r;
+  }
+
+  size_t expected_size = 1;
+  for (auto& pair : attributes) {
+    expected_size += (pair.first.size() + 1);
+    expected_size += (pair.second.size() + 1);
+  }
+
+  if (*max_len < expected_size) {
+    *max_len = expected_size;
+    return -ERANGE;
+  }
+
+  *max_len = expected_size;
+  for (auto& pair : attributes) {
+    strcpy(key_values, pair.first.c_str());
+    key_values += (strlen(key_values) + 1);
+
+    strcpy(key_values, pair.second.c_str());
+    key_values += (strlen(key_values) + 1);
+  }
+  *key_values = '\0';
+
+  return expected_size - 1;
+}
+
+extern "C" int rbd_mirror_peer_set_attributes(
+    rados_ioctx_t p, const char *uuid, const char *key_values) {
+  librados::IoCtx io_ctx;
+  librados::IoCtx::from_rados_ioctx_t(p, io_ctx);
+
+  std::map<std::string, std::string> attributes;
+  while (*key_values != '\0') {
+    const char* key = key_values;
+    key_values += strlen(key) + 1;
+    const char* value = key_values;
+    key_values += strlen(value) + 1;
```

Comment:
@dillaman So the user would need to terminate the `key_values` string with two zeroes. And `rbd_mirror_peer_get_attributes` will need an additional byte for the terminating zero (returning 1 for the empty list). It looks like we have not had functions with such behavior before? Don't you think it might be a potential source of errors?

Reply:
> (returning 1 for the empty list)

I meant if someone calls `rbd_mirror_peer_get_attributes` with zero `max_len` just to learn how large buffer it will need to store all key-values, it will return `max_len` 1 for the empty list.

Reply:
I can break it up into ```keys``` and ```values``` parameters.